gstd/gstd_http.c

```diff
@@ -0,0 +1,466 @@
+  session = GSTD_SESSION (data);
+  soup_message_headers_iter_init (&iter, msg->request_headers);
+  if (msg->request_body->length) {
+    g_print ("%s\n", msg->request_body->data);
+  }
+
+  if (msg->method == SOUP_METHOD_GET) {
+    do_get (server, msg, session);
+  } else if (msg->method == SOUP_METHOD_POST) {
+    do_post (server, msg, query, session);
+  } else if (msg->method == SOUP_METHOD_PUT) {
+    do_put (server, msg, query, session);
+  } else if (msg->method == SOUP_METHOD_DELETE) {
+    do_delete (server, msg, query, session);
+  } else {
+    soup_message_set_status (msg, SOUP_STATUS_NOT_IMPLEMENTED);
+  }
+
+  g_print ("  -> %d %s\n\n", msg->status_code, msg->reason_phrase);
+}
+
+GstdReturnCode
+gstd_http_start (GstdIpc * base, GstdSession * session)
+{
+  GError *error = NULL;
+  GstdHttp *self = GSTD_HTTP (base);
+  guint16 port = self->base_port;
+  SoupServer *server;
+  gint i;
+  
+  g_return_val_if_fail (base, GSTD_NULL_ARGUMENT);
+  g_return_val_if_fail (session, GSTD_NULL_ARGUMENT);
+
+  gstd_http_stop (base);
+
+  GST_DEBUG_OBJECT (self, "Getting HTTP address");
+  server = soup_server_new (SOUP_SERVER_SERVER_HEADER, "HTTP-Server", NULL);
+
+  for (i = 0; i < self->num_ports; i++) {
+    soup_server_listen_all (server, port + i, 0, &error);
+    if (error) {
+      goto noconnection;
+    }
+  }
+  soup_server_add_handler (server, NULL, server_callback, session, NULL);
+
+  return GSTD_EOK;
+
+noconnection:
+  {
+    GST_ERROR_OBJECT (self, "%s", error->message);
+    g_printerr ("%s\n", error->message);
+    g_error_free (error);
+    return GSTD_NO_CONNECTION;
+  }
+}
+
+gboolean
+gstd_http_init_get_option_group (GstdIpc * base, GOptionGroup ** group)
+{
+  g_return_val_if_fail (base, FALSE);
+  g_return_val_if_fail (group, FALSE);
+
+  GstdHttp *self = GSTD_HTTP (base);
+  GOptionEntry http_args[] = {
+    {"enable-http-protocol", 't', 0, G_OPTION_ARG_NONE, &base->enabled,
+        "Enable attach the server through given HTTP ports ", NULL}
+    ,
+    {"http-address", 'a', 0, G_OPTION_ARG_STRING, &self->address,
+          "Attach to the server starting through a given address (default 127.0.0.1)",
+        "http-address"}
+    ,
+    {"http-base-port", 'p', 0, G_OPTION_ARG_INT, &self->base_port,
+          "Attach to the server starting through a given port (default 5001)",
+        "http-base-port"}
+    ,
+    {"http-num-ports", 'n', 0, G_OPTION_ARG_INT, &self->num_ports,
+          "Number of ports to use starting at base-port (default 1)",
+        "http-num-ports"}
+    ,
+    {NULL}
+  };
+  GST_DEBUG_OBJECT (self, "HTTP init group callback ");
+  *group = g_option_group_new ("gstd-http", ("HTTP Options"),
+      ("Show HTTP Options"), NULL, NULL);
+
+  g_option_group_add_entries (*group, http_args);
+  return TRUE;
+}
+
+static GstdReturnCode
+gstd_http_stop (GstdIpc * base)
+{
+  g_return_val_if_fail (base, GSTD_NULL_ARGUMENT);
+  
+  GstdHttp *self = GSTD_HTTP (base);
```

Comment:
Declarations on top.

Reply:
I need to check the pointer guard before assignment.